fitter/probabilities/pulsars.py

```diff
@@ -236,29 +205,34 @@ def cluster_component(model, R, mass_bin, DM=None, ΔDM=None, DM_mdata=None, *,
             x=z_domain, y=DM_gaussian, s=0, k=3, ext=1
         )
 
-        # I think we still only want positive values for the other splines
+        # We still only want positive values for the other splines
         az_domain = np.abs(az_domain)
 
-        z1 = np.maximum(z1_spl(az_domain), z[0])
-
-        # Here we add the signs back in only for the DM splines, this allows
-        # us to select the correct side of the cluster for each az point
-        Paz_dist = (DM_los_spl(-az_signs * z1) / abs(az_der(z1))).value
-        Paz_dist <<= u.dimensionless_unscaled
-
-        outside_azt = az_domain > azt
+        Paz_dist = np.zeros_like(az_domain.value)
+        for i, a in enumerate(az_domain):
 
-        # Only use z2 if any outside_azt values exist (ensures z2_spl exists)
-        if np.any(outside_azt):
+            z_values = find_intersections(az, z, a)
 
-            z2 = z2_spl(az_domain)
+            Paz_dist[i] = np.sum(
+                (DM_los_spl(-az_signs[i] * z_values) / abs(az_der(z_values))).value,
+                axis=0,
+            )
 
-            within_bounds = outside_azt & (z2 < zt)
+        Paz_dist <<= u.dimensionless_unscaled
 
-            # Here we add the signs back in only for the DM splines
-            z2_signed = -az_signs * z2
-            Paz_dist[within_bounds] += (DM_los_spl(z2_signed[within_bounds])
-                                        / abs(az_der(z2[within_bounds]))).value
+        # This new method gives a zero in at the start of the distribution
+        # which I expect might do bad things to convolution, so I'm just
+        # going to interpolate it, this is a bit hacky but I don't see a cleaner
+        # way to do it, and this gives a nice smooth distribution
+        if np.any(Paz_dist == 0.0):
+            # get the index
+            arg_zero = np.argmin(Paz_dist)
+            # dont overflow the array
+            if 0 < arg_zero <len (Paz_dist)-1:
+                # interpolate between the two neighbouring points
+                Paz_dist[arg_zero] = (
+                    Paz_dist[arg_zero - 1] + Paz_dist[arg_zero + 1]
+                ) / 2.0
```

Comment:
Do the zeroes only happen once in a distribution? Is there a chance that multiple zeroes might appear right next to each other? If that could happen then I don't think this equation would be right.

Reply:
As far as I understand it (and this hasn't yet broken in any of my test fits), the spurious zeros should only come from that first zero that we add to the `az` domain in order to make sure the gaussian for the error on the `Pdot` measurement is defined. We do occasionally get zeros out near `zt` but I think those zeros are real. I'm not overly happy with this solution as it's pretty hacky and as you pointed out, seems a bit brittle. The fact that it's survived a few test fits makes me a bit more confident that it's good enough but if you have any ideas for how to handle this then I'd be happy to give them a shot.

Reply:
If zeros appear near `zt` and you think they're real, you probably don't want to interpolate them do you?
Without knowing what's really happening here I'm not sure what the best solution is. If you think this is working well enough then I'm sure it's fine.
Just something to be aware of going forward, I guess.

Reply:
After updating the plotting this seems to be working fine, but we should keep this in mind in case this breaks in the future.